File: playground/python/days/day02/txn_kv.py
"""Day 2 -- Challenge: Transactional Key-Value Store.

In-memory string->string store with NESTED transactions.

  set(key, value)   store a value (in the innermost open transaction,
                    or directly if none is open)
  get(key)          value currently visible for `key`, or None
  remove(key)       delete `key`; returns whether it was visible
  begin()           open a new transaction (transactions nest)
  commit()          merge the innermost transaction's changes -- writes
                    AND deletes -- into its parent transaction (or into
                    the store if it has no parent). False if no
                    transaction is open.
  rollback()        discard the innermost transaction's changes.
                    False if no transaction is open.

Reads always see the innermost state: a value set (or a key removed)
inside an open transaction is visible to get() immediately.
"""

from typing import Optional


class TxnKV:
    def __init__(self) -> None:
        # TODO: choose your own representation.
        pass

    def set(self, key: str, value: str) -> None:
        # TODO
        pass

    def get(self, key: str) -> Optional[str]:
        # TODO
        return None

    def remove(self, key: str) -> bool:
        # TODO
        return False

    def begin(self) -> None:
        # TODO
        pass

    def commit(self) -> bool:
        # TODO
        return False

    def rollback(self) -> bool:
        # TODO
        return False

File: playground/go/days/day02/txn_kv.go
package day02

// Day 2 -- Challenge: Transactional Key-Value Store.
//
// A key-value store (string -> string) with transactions. A transaction
// can be opened inside another transaction (they nest).
//
//	Set(key, value)   store a value. If a transaction is open, the change
//	                  belongs to that transaction; otherwise it is applied
//	                  directly to the store.
//	Get(key)          the value visible right now, with ok=false if the
//	                  key does not exist. Changes made inside an open
//	                  transaction are visible immediately.
//	Remove(key)       delete the key. Returns true if the key was visible
//	                  before the call. A delete made inside a transaction
//	                  also belongs to that transaction.
//	Begin()           open a new transaction.
//	Commit()          take everything the innermost transaction changed --
//	                  writes AND deletes -- and move it into the
//	                  transaction one level out (or into the store itself
//	                  if there is no outer transaction). Returns false if
//	                  no transaction is open.
//	Rollback()        throw away everything the innermost transaction
//	                  changed. Returns false if no transaction is open.

type TxnKV struct {
	// TODO: choose your own representation.
}

func NewTxnKV() *TxnKV {
	// TODO
	return &TxnKV{}
}

func (kv *TxnKV) Set(key, value string) {
	// TODO
}

func (kv *TxnKV) Get(key string) (string, bool) {
	// TODO
	return "", false
}

func (kv *TxnKV) Remove(key string) bool {
	// TODO
	return false
}

func (kv *TxnKV) Begin() {
	// TODO
}

func (kv *TxnKV) Commit() bool {
	// TODO
	return false
}

func (kv *TxnKV) Rollback() bool {
	// TODO
	return false
}
